pkg/provider/kubernetes/ingress/kubernetes.go

```diff
@@ -668,12 +669,21 @@ func (p *Provider) loadService(client Client, namespace string, backend netv1.In
 
 		protocol := getProtocol(portSpec, portName, svcConfig)
 
-		for _, addr := range subset.Addresses {
-			hostPort := net.JoinHostPort(addr.IP, strconv.Itoa(int(port)))
+		for _, endpoint := range endpointSlice.Endpoints {
+			if !(*endpoint.Conditions.Ready) {
+				continue
+			}
```

Comment:
Hello @jnoordsij!

Thanks for this contribution!

I'm wondering if filtering the endpoint condition here is introducing a breaking change in terms of discovery, WDYT?

Reply:
I did think at first when starting with this there might be a difference, but from what I've gathered it should not change anything. The endpoints with `ready=true` state on the `EndpointSlice` API should match the `addresses` on the `Endpoints` API (which also has a subsequent `notReadyAddresses` entry) and thus should return the exact same set of endpoints as before.

See also the [following note from documentation](https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/#serving):
> Although serving is almost identical to ready, it was added to prevent breaking the existing meaning of ready. It may be unexpected for existing clients if ready could be true for terminating endpoints, since historically terminating endpoints were never included in the Endpoints or EndpointSlice API to begin with. For this reason, ready is always false for terminating endpoints, and a new condition serving was added in v1.20 so that clients can track readiness for terminating pods independent of the existing semantics for ready.

Basically the new API has more information about state: rather than only doing a 'binary' separation between `ready` and not `ready`, there are now three conditions you can check on depending on your preference: `ready`, `serving` and `terminating`.

Reply:
Indeed, my first assumption was that states were not coexisting but rather a value for the state of the endpoint. Thanks for putting some thoughts and providing clarification on this 👍